Mathlib/LinearAlgebra/Matrix/FixedDetMatrices.lean

```diff
@@ -242,6 +244,45 @@ theorem induction_on {C : Δ m → Prop} {A : Δ m} (hm : m ≠ 0)
   rw [← reduce_reduceStep hc] at hA
   simpa only [reduceStep, prop_red_S hS, prop_red_T_pow hS hT] using ih hA
 
+/--The subgroup of `SL(2,ℤ)` generated by `S` and `T`. -/
+def S_T_subgroup := Subgroup.closure {S, T}
+
+lemma S_mem_S_T_subgroup : S ∈ S_T_subgroup := by
+  apply Subgroup.subset_closure
+  simp only [Set.mem_insert_iff, Set.mem_singleton_iff, true_or]
+
+lemma T_mem_S_T_subgroup : T ∈ S_T_subgroup := by
+  apply Subgroup.subset_closure
+  simp only [Set.mem_insert_iff, Set.mem_singleton_iff, or_true]
+
+lemma reps_one_id (A : FixedDetMatrix (Fin 2) ℤ 1) (a1 : A.1 1 0 = 0) (a4 : 0 < A.1 0 0)
+    (a6 : |A.1 0 1| < |(A.1 1 1)|) : A = (1 : SL(2, ℤ)) := by
+  have := Int.mul_eq_one_iff_eq_one_or_neg_one.mp (A_c_eq_zero a1)
+  ext i j
+  fin_cases i <;> fin_cases j <;> aesop
+
+lemma det_one_mem_S_T_subgroup (A : Δ 1) : A ∈ S_T_subgroup := by
+  induction A using (induction_on Int.one_ne_zero) with
+  | h0 A a1 a4 _ a6 =>
+    rw [reps_one_id A a1 a4 a6]
+    exact Subgroup.one_mem S_T_subgroup
+  | hS B hb =>
+    rw [smul_eq_mul]
+    apply Subgroup.mul_mem _ (S_mem_S_T_subgroup) (hb)
+  | hT B hb =>
+    rw [smul_eq_mul]
+    apply Subgroup.mul_mem _ (T_mem_S_T_subgroup) (hb)
+
 end IntegralFixedDetMatrices
 
 end FixedDetMatrices
+
+open MatrixGroups FixedDetMatrices
+
+section SL2Z_generators
+
+/-- `SL(2, ℤ)` is generated by `S` and `T`. -/
+lemma SpecialLinearGroup.SL2Z_generators : S_T_subgroup = ⊤ := by
+  refine (Subgroup.eq_top_iff' S_T_subgroup).mpr (fun _ => det_one_mem_S_T_subgroup _)
```

Comment:
I'm inclined to inline `S_T_subgroup`. And maybe even the entire proof.
Or do you think it will be valuable to have `S_T_subgroup` as a standalone definition later on?

Reply:
So when doing something like `Subgroup.closure_induction` one needs to give it the set, so for that its maybe useful? I'm picking `Subgroup.closure_induction` since I was using it yesteday in this context.

Reply:
Even if one is using `Subgroup.closure_induction`, I don't see why `closure {S, T} = ⊤` should be any harder to use than `S_T_subgroup = ⊤`. I've opened up #20086, but feel free to disagree.

Reply:
No that's fine, I don't have a strong preference either way!